virtwho/runner.py

```diff
@@ -1,5 +1,529 @@
+                rex = re.compile(r'(?<=: ){.*?}\n+(?=202|$)', re.S)
+                mapping_info = rex.findall(rhsm_log)[-1]
+                try:
+                    mapping_info = json.loads(mapping_info.replace('\n', ''),
+                                              strict=False)
+                except:
+                    logger.warning("Failed to run json.loads for rhsm.log")
+                    return data
+                hypervisors = mapping_info['hypervisors']
+                org_data["hypervisor_num"] = len(hypervisors)
+                for item in hypervisors:
+                    hypervisorId = item['hypervisorId']['hypervisorId']
+                    if 'name' in item.keys():
+                        hypervisor_name = item['name']
+                    else:
+                        hypervisor_name = ""
+                    facts = dict()
+                    facts['name'] = hypervisor_name
+                    facts['type'] = item['facts']['hypervisor.type']
+                    facts['version'] = item['facts']['hypervisor.version']
+                    facts['socket'] = item['facts']['cpu.cpu_socket(s)']
+                    facts['dmi'] = item['facts']['dmi.system.uuid']
+                    if 'hypervisor.cluster' in item['facts'].keys():
+                        facts['cluster'] = item['facts']['hypervisor.cluster']
+                    else:
+                        facts['cluster'] = ''
+                    guests = list()
+                    for guest in item['guestIds']:
+                        guestId = guest['guestId']
+                        guests.append(guestId)
+                        attr = dict()
+                        attr['guest_hypervisor'] = hypervisorId
+                        attr['state'] = guest['state']
+                        attr['active'] = guest['attributes']['active']
+                        attr['type'] = guest['attributes']['virtWhoType']
+                        org_data[guestId] = attr
+                    facts['guests'] = guests
+                    org_data[hypervisorId] = facts
+                data[org] = org_data
+        return data
+
+    def virtwho_print_json(self):
+        """
+        Get and return the json created by print function.
+        :return: json output
+        """
+        ret, output = self.ssh.runcmd(f"cat {self.print_json_file}")
+        if ret == 0 and output != "":
+            return output
+        else:
+            return None
+
+    def virtwho_thread_number(self):
+        """
+        Get the alive virt-who thread number.
+        :return: virt-who thread number
+        """
+        thread_num = 0
+        cmd = "ps -ef | grep virt-who -i | grep -v grep |wc -l"
+        ret, output = self.ssh.runcmd(cmd)
+        if output is not None and output != "":
+            thread_num = int(output.strip())
+        return thread_num
+
+    def run_service(self, name='virt-who', action='restart'):
+        """
+
```

Comment:
seems this empty line is redundant or miss the description...

Reply:
Done